physical/raft/raft_autopilot.go

```diff
@@ -0,0 +1,710 @@
+		MaxTrailingLogs:             1000,
+		MinQuorum:                   3,
+		ServerStabilizationTime:     10 * time.Second,
+	}
+}
+
+func (b *RaftBackend) AutopilotHCLConfig() (*AutopilotConfig, error) {
+	defaultConfig := b.defaultAutopilotConfig()
+
+	config := b.conf["autopilot"]
+	if config == "" {
+		return nil, nil
+	}
+
+	var configs []*AutopilotConfig
+	err := jsonutil.DecodeJSON([]byte(config), &configs)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed to decode autopilot config: {{err}}", err)
+	}
+	if len(configs) != 1 {
+		return nil, fmt.Errorf("expected a single block of autopilot config")
+	}
+
+	hclConfig := configs[0]
+
+	// Sanitize input
+	if hclConfig.LastContactThreshold == 0 {
+		hclConfig.LastContactThreshold = defaultConfig.LastContactThreshold
+	}
+	if hclConfig.LastContactFailureThreshold == 0 {
+		hclConfig.LastContactFailureThreshold = defaultConfig.LastContactFailureThreshold
+	}
+	if hclConfig.MaxTrailingLogs == 0 {
+		hclConfig.MaxTrailingLogs = defaultConfig.MaxTrailingLogs
+	}
+	if hclConfig.MinQuorum == 0 {
+		hclConfig.MinQuorum = defaultConfig.MinQuorum
+	}
+	if hclConfig.ServerStabilizationTime == 0 {
+		hclConfig.ServerStabilizationTime = defaultConfig.ServerStabilizationTime
+	}
+
+	return configs[0], nil
+}
+
+func (b *RaftBackend) AutopilotDisabled() bool {
+	b.l.RLock()
+	disabled := b.disableAutopilot
+	b.l.RUnlock()
+	return disabled
+}
+
+// StartAutopilot puts autopilot subsystem to work. This should only be called
+// on the active node.
+func (b *RaftBackend) StartAutopilot(ctx context.Context) {
+	b.l.RLock()
+	defer b.l.RUnlock()
+
+	if b.autopilot == nil {
+		return
+	}
+
+	b.autopilot.Start(ctx)
+	b.followerHeartbeatTrackerStopCh = make(chan struct{})
+	go b.startFollowerHeartbeatTracker()
+}
+
+// AutopilotExecutionStatus represents the current status of the autopilot background go routines
+type AutopilotExecutionStatus string
+
+const (
+	AutopilotNotRunning   AutopilotExecutionStatus = "not-running"
+	AutopilotRunning      AutopilotExecutionStatus = "running"
+	AutopilotShuttingDown AutopilotExecutionStatus = "shutting-down"
+)
+
+func autopilotStatusToStatus(status autopilot.ExecutionStatus) AutopilotExecutionStatus {
+	switch status {
+	case autopilot.Running:
+		return AutopilotRunning
+	case autopilot.ShuttingDown:
+		return AutopilotShuttingDown
+	default:
+		return AutopilotNotRunning
+	}
+}
+
+func (b *RaftBackend) startFollowerHeartbeatTracker() {
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-b.followerHeartbeatTrackerStopCh:
+			return
+		case <-ticker.C:
+			b.l.RLock()
+			if b.autopilotConfig.CleanupDeadServers && b.autopilotConfig.LastContactFailureThreshold != 0 {
+				b.followerStates.l.RLock()
+				followers := b.followerStates.followers
+				b.followerStates.l.RUnlock()
+				for id, state := range followers {
+					now := time.Now()
+					threshold := state.LastHeartbeat.Add(b.autopilotConfig.LastContactFailureThreshold)
```

Comment:
We may need extra logic here so that if the cluster was down for a day, we don't automatically mark everyone as failed immediately when we come back up.

Reply:
1) The entire cluster was down for a period.
When the active node comes up, it will reset the follower states. All computations will happen fresh and hence cluster going down shouldn't be an issue.

2) One or more followers could go down and come back up after the threshold.
2.a) Dead server cleanup is disabled.
No-op. No issues.

2.b) Dead server cleanup is enabled.
Cleanup would have already happened and the follower entry would both be already gone from raft config and followerStates in Vault. That node will need to rejoin the cluster. Everything will be fresh. No issues.

Did I miss the use-case you were thinking?

Reply:
Nope, I forgot about the `!state.LastHeartbeat.IsZero()`, so my concern wasn't warranted.  Though that does bring up another scenario to consider: suppose the active node crashes, and a new node takes over.  That new active node will have no follower state info initially, so the LastHeartbeat value it has for the former active node will be zero.  If the former active node fails to come back up, we'll never mark it dead will we?

Reply:
In that case, the new active node will never have the old active node as the follower. Vault will skip informing about that server to autopilot. Autopilot in this case considers the old active node as a stale server (not a dead server), in which case that node will be unilaterally removed by autopilot from the raft config.

It's good to have test cases for all these. Working on the same. I'm slowly adding tests. I'll get to these eventually.

Reply:
Basically, servers known to Vault but not responding can be unhealthy/dead servers. Servers unknown to Vault are stale servers.

Reply:
It looks like we populate followerStates in raftTLSRotatePhased based on the servers known to raft, so I'm not sure it's true that the new active node won't have the old active node as the follower.

Reply:
I see. I'll look into this case.

Reply:
I came to a thought that setting the LastHeartbeat to time.Now() when a node assumes active duty is enough to not encounter this issue. I've pushed up the diff with it. I'll test it out tomorrow morning.